lighthouse-core/gather/gatherers/form-elements.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * @license Copyright 2019 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+
+/**
+ *  @param {HTMLFormElement}
+ *  @return {any[]}
+ */
+/* istanbul ignore next */
+function getChildrenAInputs(formElement) {
+  if (formElement.formless == true){
+    return formElement.inputs
+  }
+
+  const inputsArray = [];
+  const childrenArray = Array.prototype.slice.call(formElement.childNodes);
+  
+  childrenArray.map(  /** @param {HTMLElement} node */ (element) => {
+    if (element.nodeName == 'INPUT' || element.nodeName == 'SELECT' || element.nodeName == 'TEXTAREA'){
+      inputAttributes = {
+        id: element.id,
+        nodeName: element.nodeName,
+        name: element.name,
+        placeholder: element.placeholder,
+        autocomplete: element.autocomplete,
+      }
+      inputsArray.push(inputAttributes);
+    }
+  });
+
+  return inputsArray;
+}
+
+/**
+ *  @param {HTMLFormElement}
+ *  @return {any[]}
+ */
+/* istanbul ignore next */
+function getChildrenLabels(formElement) {
+  if (formElement.formless == true){
+    return formElement.labels
+  }
+  const childrenArray = Array.prototype.slice.call(formElement.childNodes);
+  const labelsArray = [];
+  childrenArray.map(/** @param {HTMLElement} node */ (element) => {
+    if (element.nodeName == 'LABEL'){
+      labelAttributes = {
+        id: element.id,
+        nodeName: element.nodeName,
+        name: element.name,
+        for: element.for,
+      }
+      labelsArray.push(labelAttributes);
+    }
+  });
+
+  return labelsArray;
+}
+
+
+/**
+ * @return {LH.Artifacts['FormElements']}
+ */
+/* istanbul ignore next */
+function collectFormElements() {
+  // @ts-ignore - put into scope via stringification
+  const inputElements = getElementsInDocument('input'); // eslint-disable-line no-undef
+  const selectElements = getElementsInDocument('select'); // eslint-disable-line no-undef
+  const textareaElements = getElementsInDocument('textarea'); // eslint-disable-line no-undef
+  const labelElements = getElementsInDocument('label'); // eslint-disable-line no-undef
+  const formElements = getElementsInDocument('form'); // eslint-disable-line no-undef
+  const formChildren = inputElements.concat(selectElements, textareaElements, labelElements);
+
+  const formless = {
+    formless: true,
```

Comment:
What is this field used for since it seems to always be set to true?

Reply:
used in the getChildrenInputs and getChildrenLabels functions

Reply:
This can be replaced by checking for the existence of `.form` on these objects (your L114-117).

Reply:
@connorjclark for the form object?

Reply:
right, so instead of checking `formless` in those two fns you mentioned you can just do `if (formElement.form) ...`